### ISSUE TITLE
Pooling support for 5D or ND-Blobs

### ISSUE THREAD

**Issue by henzler** [BVLC/caffe#5048]
It is possible to have ND-Blobs but when it comes to pooling caffe does not support Blobs of any other dimension but 4D.

> Check failed: 4 == bottom[0]->num_axes() (4 vs. 5) Input must have 4 axes, corresponding to (num, channels, height, width)

Can you add ND-pooling support?

**Comment by shaibagon** [comment]
I would like to "up vote" this request. Caffe moved to support ND arrays long ago. This transition included a change to `"Blob"` data structure and to many layers, especially to `"Convolution"` layer.
However, `"Pooling"` layer currently is restricted to "legacy" 4D blobs only, which really restricts the deployment of caffe to more challenging domain, e.g., video processing, 3D medical scans etc.

Please see [this SO thread](http://stackoverflow.com/q/40888514/1714410) related to this issue.

**Comment by naibaf7** [comment]
@thigitogatiamas @shaibagon 
Full ND support for Pooling is available for quite some time for both CUDA and OpenCL in the OpenCL Caffe branch. Not sure when the transition will be concluded for the master branch.

**Comment by bhack** [comment]
...And then libdnn porting right? ;)

**Comment by naibaf7** [comment]
@bhack 
LibDNN already has a ND-pooling engine superior to that of Caffe. I'm currently fixing some corner cases and improving it, so that it can be signed off for the standalone LibDNN :); all according to the timeplan we communicated earlier :)

**Comment by bhack** [comment]
Yes but can we plan to change names at some point? Libdnn it is becoming ambiguous

**Comment by naibaf7** [comment]
@bhack Change the name of what where? I mean within Caffe, LibDNN is identical to the standalone. The reason I can't use the standalone in Caffe is because standalone exports some Caffe functionality to use with other frameworks...

**Comment by bhack** [comment]
Yes the problem is that we use the same name for  an artifact that is conceptually the same but for tuning requirements are not in sync cause there is need to duplicate functionalites already in Caffe to let execute the tuning and all the other bootstrap stuffs.

**Comment by shaibagon** [comment]
@bhack @naibaf7  I lost you here... Can you give an estimate to a commoner when is ND pooling expected to be part of caffe/master?

Thank you very much for all your efforts developing and driving caffe forward!

**Comment by bhack** [comment]
@shaibagon Nevermind. I've have created some noise in the thread talking about https://github.com/naibaf7/libdnn that is the standalone version of what you are waiting in caffe master.

**Comment by naibaf7** [comment]
@shaibagon 
Yes, sorry, ND support is here: https://github.com/BVLC/caffe/tree/opencl

**Comment by sajjo79** [comment]
Hi,
I am trying to figure out the 3D convolution and pooling in caffer master branch. Is there any such support available?
